csrc/host_ir/jit.cpp

```diff
@@ -68,185 +78,248 @@ T throwIfError(llvm::Expected<T>&& E) {
+        builder.CreateStore(tv2atenMap[tv], aten_tensor_ptr);
+      }
+    }
+    llvm::Value* result = builder.CreateBitCast(aten_tensor_array_ptr, getType(JitDataType::void_array_ptr, ctx));
+    builder.CreateRet(result);
+
+    const bool debug_print = isDebugDumpEnabled(DebugDumpOption::HostIrJit);
+    if (debug_print) {
+      llvm::outs() << "=== LLVM IR Generation for Main Function Outputs ===\n";
+      mod->print(llvm::outs(), nullptr);
+    }
+}
+
+
+void compileFunctionDeclarations(
+    llvm::Module* mod,
+    llvm::LLVMContext& ctx) {
+
+  // get the types
+  auto* void_ptr_type = getType(JitDataType::void_ptr, ctx);
+  auto* void_array_ptr_type = getType(JitDataType::void_array_ptr, ctx);
+  auto* int64_t_type = getType(JitDataType::int64_t, ctx);
+
+  // tensor_size function: int64_t tensor_size(at::Tensor* tensor, int64_t dim)
+  llvm::FunctionType* tensor_size_type = llvm::FunctionType::get(int64_t_type, {void_ptr_type, int64_t_type}, false);
+  llvm::Function::Create(tensor_size_type, llvm::Function::ExternalLinkage, tensor_size_func_name, mod);
+
+  llvm::FunctionType* allocate_tensor_type = llvm::FunctionType::get(void_ptr_type, {}, false);
+  llvm::Function::Create(allocate_tensor_type, llvm::Function::ExternalLinkage, allocate_tensor_func_name, mod);
+
+  // set_tensor function: void set_tensor(at::Tensor* tensor, at::Tensor* other_tensor)
+  llvm::FunctionType* set_tensor_type = llvm::FunctionType::get(llvm::Type::getVoidTy(ctx), {void_ptr_type, void_ptr_type}, false);
+  llvm::Function::Create(set_tensor_type, llvm::Function::ExternalLinkage, set_tensor_func_name, mod);
+
+  // main function: at::Tensor** main(at::Tensor** input_tensors)
+  llvm::FunctionType* main_type = llvm::FunctionType::get(void_array_ptr_type, {void_array_ptr_type}, false);
+  llvm::Function::Create(main_type, llvm::Function::ExternalLinkage, main_func_name, mod);
+
+}
+
+void compile(HostIrJitImpl* pimpl) {
   FUSER_PERF_SCOPE("HostIrJit::compile");
   NVF_ERROR(
-      container != nullptr,
+      pimpl->container != nullptr,
       "container is nullptr during host ir JIT compilation");
   auto ctx = std::make_unique<llvm::LLVMContext>();
   auto mod = std::make_unique<llvm::Module>("host_ir_jit_module", *ctx);
-  std::vector<kir::Allocate*> allocate_exprs;
-  for (auto* expr : container->topLevelExprs()) {
-    if (auto* allocate = dynamic_cast<kir::Allocate*>(expr)) {
-      createAndInsertAllocationFunction(allocate, mod.get());
-      allocate_exprs.push_back(allocate);
-    } else if (auto* for_loop = dynamic_cast<ForLoop*>(expr)) {
-      for (auto* expr : for_loop->body().exprs()) {
-        if (auto* allocate = dynamic_cast<kir::Allocate*>(expr)) {
-          createAndInsertAllocationFunction(allocate, mod.get());
-          allocate_exprs.push_back(allocate);
-        }
-      }
+  llvm::IRBuilder<> builder(*ctx);
+  std::unordered_map<Val*, llvm::Value*> val2llvmMap;
+  std::unordered_map<TensorView*, llvm::Value*> tv2atenMap;
```

Comment:
I'm confused by `tv2atenMap`. The value type is not an ATen type. 

Also, can val_to_llvm subsume tv2atenMap? TensorView is a subclass of Val.